From 1ae99ae2bcf063297c966fd9cb8782fbebd0ab50 Mon Sep 17 00:00:00 2001
From: Chen Meng J <meng.j.chen@intel.com>
Date: Tue, 11 Sep 2018 10:24:50 +0800
Subject: [PATCH 51/54] media: intel-ipu4: ox03a10: more exposure gain CID
 support

Message for Open Source:
more exposure gain CID support

Message for Internal:
[Feature] sensor ox03a10 driver enabling.
add CID of exposure, digital gain, analog gain,
for channels of hcg lcg vs.

Change-Id: I982dac483a8d5866fe67c05258ba639870b24f96
Tracked-On: #JIIAP-573
Signed-off-by: Chen Meng J <meng.j.chen@intel.com>
Signed-off-by: Meng Wei <wei.meng@intel.com>
---
 .../i2c/crlmodule/crl_ox03a10_configuration.h      | 150 +++++++++++++--------
 1 file changed, 93 insertions(+), 57 deletions(-)

diff --git a/drivers/media/i2c/crlmodule/crl_ox03a10_configuration.h b/drivers/media/i2c/crlmodule/crl_ox03a10_configuration.h
index 43bde8c..4b44afe 100644
--- a/drivers/media/i2c/crlmodule/crl_ox03a10_configuration.h
+++ b/drivers/media/i2c/crlmodule/crl_ox03a10_configuration.h
@@ -1236,23 +1236,59 @@ struct crl_v4l2_ctrl ox03a10_v4l2_ctrls[] = {
 		.sd_type = CRL_SUBDEV_TYPE_PIXEL_ARRAY,
 		.op_type = CRL_V4L2_CTRL_SET_OP,
 		.context = SENSOR_POWERED_ON,
-		.name = "CRL_CID_EXPOSURE_MODE",
-		.ctrl_id = CRL_CID_EXPOSURE_MODE,
+		.name = "DCG exposure",
+		.ctrl_id =  V4L2_CID_EXPOSURE,
 		.type = CRL_V4L2_CTRL_TYPE_CUSTOM,
-		.data.std_data.min = 0,
-		.data.std_data.max = 1,
+		.data.std_data.min = 1,
+		.data.std_data.max = 1280,
 		.data.std_data.step = 1,
-		.data.std_data.def = 0,
+		.data.std_data.def = 0x40,
 		.flags = V4L2_CTRL_FLAG_UPDATE,
-		.impact = CRL_IMPACTS_MODE_SELECTION,
+		.impact = CRL_IMPACTS_NO_IMPACT,
+		.regs_items = ARRAY_SIZE(ox03a10_dcg_exposure_coarse),
+		.regs = ox03a10_dcg_exposure_coarse,
 		.v4l2_type = V4L2_CTRL_TYPE_INTEGER,
 	},
 	{
 		.sd_type = CRL_SUBDEV_TYPE_PIXEL_ARRAY,
 		.op_type = CRL_V4L2_CTRL_SET_OP,
 		.context = SENSOR_POWERED_ON,
-		.name = "HCG ANALOG GAIN 5.6",
-		.ctrl_id = CRL_CID_ANALOG_GAIN_L,
+		.name = "VS exposure",
+		.ctrl_id =  CRL_CID_EXPOSURE_SHS1,
+		.type = CRL_V4L2_CTRL_TYPE_CUSTOM,
+		.data.std_data.min = 1,
+		.data.std_data.max = 1280,
+		.data.std_data.step = 1,
+		.data.std_data.def = 0x40,
+		.flags = V4L2_CTRL_FLAG_UPDATE,
+		.impact = CRL_IMPACTS_NO_IMPACT,
+		.regs_items = ARRAY_SIZE(ox03a10_vs_exposure_coarse),
+		.regs = ox03a10_vs_exposure_coarse,
+		.v4l2_type = V4L2_CTRL_TYPE_INTEGER,
+	},
+	{
+		.sd_type = CRL_SUBDEV_TYPE_PIXEL_ARRAY,
+		.op_type = CRL_V4L2_CTRL_SET_OP,
+		.context = SENSOR_POWERED_ON,
+		.name = "LCG digital gain",
+		.ctrl_id = V4L2_CID_GAIN,
+		.type = CRL_V4L2_CTRL_TYPE_CUSTOM,
+		.data.std_data.min = 0x400,
+		.data.std_data.max = 0x3FFF,
+		.data.std_data.step = 1,
+		.data.std_data.def = 0x400,
+		.flags = V4L2_CTRL_FLAG_UPDATE,
+		.impact = CRL_IMPACTS_NO_IMPACT,
+		.regs_items = ARRAY_SIZE(ox03a10_lcg_digital_gain),
+		.regs = ox03a10_hcg_digital_gain,
+		.v4l2_type = V4L2_CTRL_TYPE_INTEGER,
+	},
+	{
+		.sd_type = CRL_SUBDEV_TYPE_PIXEL_ARRAY,
+		.op_type = CRL_V4L2_CTRL_SET_OP,
+		.context = SENSOR_POWERED_ON,
+		.name = "LCG analog gain",
+		.ctrl_id =  V4L2_CID_ANALOGUE_GAIN,
 		.type = CRL_V4L2_CTRL_TYPE_CUSTOM,
 		.data.std_data.min = 0x40,
 		.data.std_data.max = 0x400,
@@ -1260,7 +1296,7 @@ struct crl_v4l2_ctrl ox03a10_v4l2_ctrls[] = {
 		.data.std_data.def = 0x40,
 		.flags = V4L2_CTRL_FLAG_UPDATE,
 		.impact = CRL_IMPACTS_NO_IMPACT,
-		.regs_items = ARRAY_SIZE(ox03a10_hcg_real_gain),
+		.regs_items = ARRAY_SIZE(ox03a10_lcg_real_gain),
 		.regs = ox03a10_hcg_real_gain,
 		.v4l2_type = V4L2_CTRL_TYPE_INTEGER,
 	},
@@ -1268,8 +1304,8 @@ struct crl_v4l2_ctrl ox03a10_v4l2_ctrls[] = {
 		.sd_type = CRL_SUBDEV_TYPE_PIXEL_ARRAY,
 		.op_type = CRL_V4L2_CTRL_SET_OP,
 		.context = SENSOR_POWERED_ON,
-		.name = "HCG DITIGAL GAIN 4.10",
-		.ctrl_id = CRL_CID_DIGITAL_GAIN_L,
+		.name = "LCG digital gain",
+		.ctrl_id = V4L2_CID_DIGITAL_GAIN,
 		.type = CRL_V4L2_CTRL_TYPE_CUSTOM,
 		.data.std_data.min = 0x400,
 		.data.std_data.max = 0x3FFF,
@@ -1277,7 +1313,7 @@ struct crl_v4l2_ctrl ox03a10_v4l2_ctrls[] = {
 		.data.std_data.def = 0x400,
 		.flags = V4L2_CTRL_FLAG_UPDATE,
 		.impact = CRL_IMPACTS_NO_IMPACT,
-		.regs_items = ARRAY_SIZE(ox03a10_hcg_digital_gain),
+		.regs_items = ARRAY_SIZE(ox03a10_lcg_digital_gain),
 		.regs = ox03a10_hcg_digital_gain,
 		.v4l2_type = V4L2_CTRL_TYPE_INTEGER,
 	},
@@ -1285,42 +1321,44 @@ struct crl_v4l2_ctrl ox03a10_v4l2_ctrls[] = {
 		.sd_type = CRL_SUBDEV_TYPE_PIXEL_ARRAY,
 		.op_type = CRL_V4L2_CTRL_SET_OP,
 		.context = SENSOR_POWERED_ON,
-		.name = "DCG exposure in multi exposure",
-		.ctrl_id =  CRL_CID_EXPOSURE_SHS1,
+		.name = "HCG analog gain",
+		.ctrl_id = CRL_CID_ANALOG_GAIN_S,
 		.type = CRL_V4L2_CTRL_TYPE_CUSTOM,
-		.data.std_data.min = 1,
-		.data.std_data.max = 1280,
+		.data.std_data.min = 0x40,
+		.data.std_data.max = 0x400,
 		.data.std_data.step = 1,
 		.data.std_data.def = 0x40,
 		.flags = V4L2_CTRL_FLAG_UPDATE,
 		.impact = CRL_IMPACTS_NO_IMPACT,
-		.regs_items = ARRAY_SIZE(ox03a10_dcg_exposure_coarse),
-		.regs = ox03a10_dcg_exposure_coarse,
+		.regs_items = ARRAY_SIZE(ox03a10_hcg_real_gain),
+		.regs = ox03a10_hcg_real_gain,
 		.v4l2_type = V4L2_CTRL_TYPE_INTEGER,
 	},
 	{
 		.sd_type = CRL_SUBDEV_TYPE_PIXEL_ARRAY,
 		.op_type = CRL_V4L2_CTRL_SET_OP,
 		.context = SENSOR_POWERED_ON,
-		.name = "DCG exposure",
-		.ctrl_id =  V4L2_CID_EXPOSURE,
+		.name = "HCG digital gain",
+		.ctrl_id = CRL_CID_DIGITAL_GAIN_S,
 		.type = CRL_V4L2_CTRL_TYPE_CUSTOM,
-		.data.std_data.min = 1,
-		.data.std_data.max = 1280,
+		.data.std_data.min = 0x400,
+		.data.std_data.max = 0x3FFF,
 		.data.std_data.step = 1,
-		.data.std_data.def = 0x40,
+		.data.std_data.def = 0x400,
 		.flags = V4L2_CTRL_FLAG_UPDATE,
 		.impact = CRL_IMPACTS_NO_IMPACT,
-		.regs_items = ARRAY_SIZE(ox03a10_dcg_exposure_coarse),
-		.regs = ox03a10_dcg_exposure_coarse,
+		.regs_items = ARRAY_SIZE(ox03a10_hcg_digital_gain),
+		.regs = ox03a10_hcg_digital_gain,
 		.v4l2_type = V4L2_CTRL_TYPE_INTEGER,
 	},
+
+
 	{
 		.sd_type = CRL_SUBDEV_TYPE_PIXEL_ARRAY,
 		.op_type = CRL_V4L2_CTRL_SET_OP,
 		.context = SENSOR_POWERED_ON,
-		.name = "LCG ANALOG GAIN",
-		.ctrl_id = CRL_CID_ANALOG_GAIN_S,
+		.name = "LCG analog gain",
+		.ctrl_id = CRL_CID_ANALOG_GAIN_L,
 		.type = CRL_V4L2_CTRL_TYPE_CUSTOM,
 		.data.std_data.min = 0x40,
 		.data.std_data.max = 0x400,
@@ -1336,8 +1374,8 @@ struct crl_v4l2_ctrl ox03a10_v4l2_ctrls[] = {
 		.sd_type = CRL_SUBDEV_TYPE_PIXEL_ARRAY,
 		.op_type = CRL_V4L2_CTRL_SET_OP,
 		.context = SENSOR_POWERED_ON,
-		.name = "LCG DITIGAL GAIN",
-		.ctrl_id = CRL_CID_DIGITAL_GAIN_S,
+		.name = "LCG digital gain",
+		.ctrl_id = CRL_CID_DIGITAL_GAIN_L,
 		.type = CRL_V4L2_CTRL_TYPE_CUSTOM,
 		.data.std_data.min = 0x400,
 		.data.std_data.max = 0x3FFF,
@@ -1353,7 +1391,7 @@ struct crl_v4l2_ctrl ox03a10_v4l2_ctrls[] = {
 		.sd_type = CRL_SUBDEV_TYPE_PIXEL_ARRAY,
 		.op_type = CRL_V4L2_CTRL_SET_OP,
 		.context = SENSOR_POWERED_ON,
-		.name = "VS ANALOG GAIN",
+		.name = "VS analog gain",
 		.ctrl_id =  CRL_CID_ANALOG_GAIN_VS,
 		.type = CRL_V4L2_CTRL_TYPE_CUSTOM,
 		.data.std_data.min = 0x40,
@@ -1370,7 +1408,7 @@ struct crl_v4l2_ctrl ox03a10_v4l2_ctrls[] = {
 		.sd_type = CRL_SUBDEV_TYPE_PIXEL_ARRAY,
 		.op_type = CRL_V4L2_CTRL_SET_OP,
 		.context = SENSOR_POWERED_ON,
-		.name = "VS DITIGAL GAIN",
+		.name = "VS digital gain",
 		.ctrl_id = CRL_CID_DIGITAL_GAIN_VS,
 		.type = CRL_V4L2_CTRL_TYPE_CUSTOM,
 		.data.std_data.min = 0x400,
@@ -1387,17 +1425,15 @@ struct crl_v4l2_ctrl ox03a10_v4l2_ctrls[] = {
 		.sd_type = CRL_SUBDEV_TYPE_PIXEL_ARRAY,
 		.op_type = CRL_V4L2_CTRL_SET_OP,
 		.context = SENSOR_POWERED_ON,
-		.name = "VS exposure",
-		.ctrl_id =  CRL_CID_EXPOSURE_SHS2,
+		.name = "CRL_CID_EXPOSURE_MODE",
+		.ctrl_id = CRL_CID_EXPOSURE_MODE,
 		.type = CRL_V4L2_CTRL_TYPE_CUSTOM,
-		.data.std_data.min = 1,
-		.data.std_data.max = 1280,
+		.data.std_data.min = 0,
+		.data.std_data.max = 1,
 		.data.std_data.step = 1,
-		.data.std_data.def = 0x40,
+		.data.std_data.def = 0,
 		.flags = V4L2_CTRL_FLAG_UPDATE,
-		.impact = CRL_IMPACTS_NO_IMPACT,
-		.regs_items = ARRAY_SIZE(ox03a10_vs_exposure_coarse),
-		.regs = ox03a10_vs_exposure_coarse,
+		.impact = CRL_IMPACTS_MODE_SELECTION,
 		.v4l2_type = V4L2_CTRL_TYPE_INTEGER,
 	},
 	{
@@ -1419,24 +1455,6 @@ struct crl_v4l2_ctrl ox03a10_v4l2_ctrls[] = {
 		.dep_ctrls = 0,
 	},
 	{
-		.sd_type = CRL_SUBDEV_TYPE_BINNER,
-		.op_type = CRL_V4L2_CTRL_GET_OP,
-		.context = SENSOR_POWERED_ON,
-		.ctrl_id = V4L2_CID_PIXEL_RATE,
-		.name = "V4L2_CID_PIXEL_RATE_CSI",
-		.type = CRL_V4L2_CTRL_TYPE_INTEGER,
-		.data.std_data.min = 0,
-		.data.std_data.max = INT_MAX,
-		.data.std_data.step = 1,
-		.data.std_data.def = 0,
-		.flags = 0,
-		.impact = CRL_IMPACTS_NO_IMPACT,
-		.regs_items = 0,
-		.regs = 0,
-		.dep_items = 0,
-		.dep_ctrls = 0,
-	},
-	{
 		.sd_type = CRL_SUBDEV_TYPE_PIXEL_ARRAY,
 		.op_type = CRL_V4L2_CTRL_SET_OP,
 		.context = SENSOR_POWERED_ON,
@@ -1478,6 +1496,24 @@ struct crl_v4l2_ctrl ox03a10_v4l2_ctrls[] = {
 	},
 	{
 		.sd_type = CRL_SUBDEV_TYPE_BINNER,
+		.op_type = CRL_V4L2_CTRL_GET_OP,
+		.context = SENSOR_POWERED_ON,
+		.ctrl_id = V4L2_CID_PIXEL_RATE,
+		.name = "V4L2_CID_PIXEL_RATE_CSI",
+		.type = CRL_V4L2_CTRL_TYPE_INTEGER,
+		.data.std_data.min = 0,
+		.data.std_data.max = INT_MAX,
+		.data.std_data.step = 1,
+		.data.std_data.def = 0,
+		.flags = 0,
+		.impact = CRL_IMPACTS_NO_IMPACT,
+		.regs_items = 0,
+		.regs = 0,
+		.dep_items = 0,
+		.dep_ctrls = 0,
+	},
+	{
+		.sd_type = CRL_SUBDEV_TYPE_BINNER,
 		.op_type = CRL_V4L2_CTRL_SET_OP,
 		.context = SENSOR_IDLE,
 		.ctrl_id = V4L2_CID_LINK_FREQ,
-- 
1.9.1

